### PR TITLE
Add TimeRangeConstraint Representation Invariant

### PR DESCRIPTION
**Pull request by plypaul**
As per title.

### CODE REVIEW

File: metricflow/constraints/time_constraint.py [hunk 1]
```diff
@@ -17,6 +17,20 @@ class TimeRangeConstraint(SerializableDataclass):
     start_time: datetime.datetime
     end_time: datetime.datetime
 
+    def __post_init__(self) -> None:  # noqa: D
+        if self.start_time > self.end_time:
+            raise RuntimeError(
+                f"start_time must not be > end_time. start_time={self.start_time} end_time={self.end_time}"
+            )
+
+        if self.start_time < TimeRangeConstraint.ALL_TIME_BEGIN():
```
Comment (tlento) [inline metricflow/constraints/time_constraint.py:26]: Shall we switch this to epoch start instead of the year 2000? Seems more in line with other computer date systems.

File: metricflow/constraints/time_constraint.py [hunk 2]
```diff
@@ -17,6 +17,20 @@ class TimeRangeConstraint(SerializableDataclass):
...
+            raise RuntimeError(
+                f"start_time={self.start_time} exceeds the limits of {TimeRangeConstraint.ALL_TIME_BEGIN()}"
+            )
+
+        if self.end_time > TimeRangeConstraint.ALL_TIME_END():
+            raise RuntimeError(f"end_time={self.end_time} exceeds the limits of {TimeRangeConstraint.ALL_TIME_END()}")
```
Comment (tlento) [inline metricflow/constraints/time_constraint.py:32]: 2040 isn't too far away but I guess that's alright.

Out of historical instance, why is this here? And why is it a fixed constant? Shouldn't we make it slide forward based on current date?

**Review by tlento — COMMENTED** [review]
Conceptually great, needs a fix for that test failure, and I think the right thing to do is update our range boundaries.

File: metricflow/constraints/time_constraint.py (hunk 2 above)
Comment (plypaul) [reply metricflow/constraints/time_constraint.py:32]: So the limits of the `TimeRangeConstraint` are used for the time spine table. We could increase this, but that would cause the size of the time spine table to be larger, which adds a bit of overhead when creating it.

Generally prefer not to make it sliding to get better consistency + easier debugging.

Comment (tlento) [reply metricflow/constraints/time_constraint.py:32]: > the limits of the TimeRangeConstraint are used for the time spine table

Ah, makes sense!

File: metricflow/test/time/test_time_granularity_solver.py [hunk 3]
```diff
@@ -223,10 +223,7 @@ def test_adjusted_time_constraint_at_boundaries(
     """Tests the time range adjustment behavior at the min / max boundaries"""
     assert (
         time_granularity_solver.adjust_time_range_to_granularity(
-            time_range_constraint=TimeRangeConstraint(
-                start_time=TimeRangeConstraint.all_time().start_time + datetime.timedelta(days=-30),
-                end_time=TimeRangeConstraint.all_time().end_time + datetime.timedelta(days=30),
```
Comment (tlento) [inline metricflow/test/time/test_time_granularity_solver.py:228]: LOL. The reason this failed is because the timedelta signs were backwards.

**Review by tlento — APPROVED** [review]
LGTM!
